runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -75,59 +80,55 @@ public int size() {
      * with their snapshot+index id (also a long) which represents a global comparable ordering.
      * This class encapsulates these two longs into one Id and add rules on comparability.
      */
+    @EqualsAndHashCode
     public static class CorfuRecordId implements Comparable<CorfuRecordId> {
-        private final UUID id;
-        public CorfuRecordId(long ordering, long uniqueId) {
-            this.id = new UUID(ordering, uniqueId);
-        }
+        @Setter
+        @Getter
+        private long epoch;
 
-        /**
-         * @return Return only the unique part of the id without the ordering
-         */
-        public long getEntryId() {
-            return id.getLeastSignificantBits();
+        @Setter
+        @Getter
+        private long sequence;
+
+        @Getter
+        private long entryId;
+
+        public CorfuRecordId(long epoch, long sequence, long entryId) {
+            this.epoch = epoch;
+            this.sequence = sequence;
+            this.entryId = entryId;
         }
 
         /**
-         * @return Return only the ordering part of the entry without the id.
+         * @return Pack CorfuRecordId into a 16 byte UUID
+         * Q: Is it safe?
+         * At the rate of 2ms per transaction, sequence will take  69 years to rollover.
+         * At the rate of 1 cluster reconfig per minute, epoch will take 31 years to rollover.
          */
-        public long getOrdering() {
-            return id.getMostSignificantBits();
+        public UUID asUUID() {
+            final int BITS_FOR_SEQUENCE = 40;
+            return new UUID(epoch<<BITS_FOR_SEQUENCE | sequence, entryId);
         }
 
         /**
          * It's NOT ok to compare two objects if their ordering metadata is dissimilar.
          * @param o object to compare against.
-         * @throws IllegalArgumentException if the two Ids are not comparable.
          * @return results of comparison.
          */
         @Override
         public int compareTo(CorfuRecordId o) {
-            if (this.id.getMostSignificantBits() == 0 && o.id.getMostSignificantBits() != 0) {
-                throw new IllegalArgumentException(
-                        "Incompatible CorfuRecordId comparison: ordering unavailable");
-            }
-            if (this.id.getMostSignificantBits() !=0 && o.id.getMostSignificantBits() == 0) {
-               throw new IllegalArgumentException(
-                       "Incompatible CorfuRecordId comparison: order of compared object unknown");
-            }
-            if (this.id.getLeastSignificantBits() == o.id.getLeastSignificantBits()) {
-                return 0;
+            if ((this.sequence != Address.NON_ADDRESS && o.sequence == Address.NON_ADDRESS) ||
+                    (this.sequence == Address.NON_ADDRESS && o.sequence != Address.NON_ADDRESS)) {
+                throw new IllegalStateException("CorfuQueue cannot mix transactional and non-transactional types");
```

Comment:
Looks like you are still using 0, 0 for non-transactional entries (the else branch in enqueue) ?

Reply:
exactly, Queue is still usable if all enqueue() are non-transactional (unit tests only at this point)

Reply:
Sorry good catch

Reply:
@hisundar Is this fixed?

Reply:
yes I removed the check as you suggested.